### ISSUE TITLE
Conditionals can cause attribute type narrowing to fail

### ISSUE THREAD

**Issue by GBeauregard** [facebook/pyre-check#575]
# Pyre Bug

**Bug description**
In certain cases conditionals can cause class attribute type narrowing to fail.

**Reproduction steps**
```python
import pathlib
from typing import Optional

class ContainsPath:
    mypath: Optional[pathlib.Path]

    def __init__(self) -> None:
        self.mypath = None

def func(var: ContainsPath) -> None:
    if var.mypath is None:
        pass
    elif var.mypath.is_file():
        reveal_type(var.mypath)
        var.mypath.read_text()  # false positive error
```

**Expected behavior**
The above code should type check and the type should be revealed as `Path` instead of `Optional[Path]`. Both mypy and pyright can handle this code snippet.

**Logs**

```console
ƛ No watchman binary found.
To enable pyre incremental, you can install watchman: https://facebook.github.io/watchman/docs/install
ƛ Defaulting to non-incremental check.
ƛ Using virtual environment site-packages in search path...
ƛ Found 2 type errors!
test/min.py:14:8 Revealed type [-1]: Revealed type for `var.mypath` is `Optional[pathlib.Path]`.
test/min.py:15:8 Undefined attribute [16]: `Optional` has no attribute `read_text`.
```

**Additional context**
Client version: 0.9.10
Binary version: d418f5bb10fdee9648eeea3834c8ca56944bd2ce

This could be seen as intentional in the same way that https://github.com/facebook/pyre-check/issues/576 is

**Comment by GBeauregard** [comment]
Closing as intentional https://pyre-check.org/docs/errors/#optional-attributes
